### PR TITLE
Nio 1141 poll

### PR DESCRIPTION
**Pull request by tlugger**
A lot of the logic for pulling and updating running instance configurations has been moved over to the `manager` of this component. The handler now just collects relevant information through the PUT request and calls the relevant manager function.

### CODE REVIEW

File: manager.py [hunk 1]
```diff
@@ -55,43 +71,121 @@ def configure(self, context):
...
             "configuration", "delete_missing", fallback=False)
 
-        # TODO in NIO-1141: Pull config settings for autonomy
+        # fetch instance specific settings
+        default = Persistence().load("api_key", default=None)
+        self.api_key = NIOEnvironment.get_variable('API_KEY', 
+                                                   default=default)
+
+        # config autonomy spcific settings
+        self._poll = Settings.getboolean("configuration", "config_polling",
```
Comment (f1401martin) [inline manager.py]: I believe we normally don't require a boolean setting for interval-like configurations, meaning a poll interval equal to 0 means no polling.

File: manager.py [hunk 2]
```diff
@@ -55,43 +71,121 @@ def configure(self, context):
...
-
-        self._instance_id = Settings.get("configuration", "instance_id")
-       
+        self.config_api_url_prefix = \
+            Settings.get("configuration", "config_api_url_prefix",
+                fallback="https://api.nio.works/v1/instance_configurations")
+
+        setting_config_id = Settings.get("configuration", "config_id",
+            fallback=None)
+        self.config_id = Persistence().load("configuration_id",
```
Comment (f1401martin) [inline manager.py]: I am being picky here but if most settings calling for a `config-like` prefix use `config`, probably this one should too.

File: manager.py [hunk 3]
```diff
@@ -55,43 +71,121 @@ def configure(self, context):
...
+                                            default=setting_config_id)
+        setting_config_version_id = \
+            Settings.get("configuration", "config_version_id",
+                fallback=None)
+        self.config_version_id = Persistence().\
```
Comment (f1401martin) [inline manager.py]: This reads it from Persistence and saves it back, it should assign to _config_version_id

File: manager.py (hunk 2 above)
Comment (f1401martin) [inline manager.py]: This reads it from Persistence and saves it back, it should assign to _config_id

File: manager.py [hunk 4]
```diff
@@ -55,43 +71,121 @@ def configure(self, context):
...
+        # TODO: In NIO-1142, this should first get the newest config version 
+        # from the product api before triggering an update
+        url = "{}/{}/versions/{}".format(self.config_api_url_prefix,
+                                         self.config_id,
+                                         self.config_version_id)
+        result = self.update_configuration(url)
+        self.logger.info("Configuration was updated, {}".format(result))
+
+    def update_configuration(self, url):
+        configuration = self._proxy.load_configuration(url, self.api_key) or {}
```
Comment (f1401martin) [inline manager.py]: replace `_proxy` with `_api_proxy`

File: manager.py [hunk 5]
```diff
@@ -3,16 +3,23 @@
    Configuration Manager
 
 """
+import json
```
Comment (f1401martin) [inline manager.py]: not used

File: manager.py [hunk 6]
```diff
@@ -3,16 +3,23 @@
...
 from nio.util.versioning.dependency import DependsOn
 from nio import discoverable
-from nio.modules.settings import Settings
 
+from nio.modules.settings import Settings
 from nio.modules.persistence import Persistence
+from nio.modules.scheduler.job import Job
+
 from niocore.util.environment import NIOEnvironment
+from niocore.configuration import CfgType
```
Comment (f1401martin) [inline manager.py]: not used

File: manager.py [hunk 7]
```diff
@@ -39,14 +53,16 @@ def configure(self, context):
...
-        Fetches settings and instantiates ConfigHandler
+        Fetches settings
 
         Args:
             context (CoreContext): component initialization context
 
         """
 
         super().configure(context)
+        self._configuration_manager = \
```
Comment (f1401martin) [inline manager.py]: This is repeated below

File: manager.py [hunk 8]
```diff
@@ -55,43 +71,121 @@ def configure(self, context):
         self.logger.info('REST Manager set to {}'.format(self._rest_manager))
```
Comment (f1401martin) [inline manager.py]: Might want to remove this, it isn't that helpful

File: manager.py [hunk 9]
```diff
@@ -55,43 +71,121 @@ def configure(self, context):
...
+                                                   default=default)
+
+        # config autonomy spcific settings
+        self._poll = Settings.getboolean("configuration", "config_polling",
+                                         fallback=False)
+                                         
+        self._poll_interval = Settings.getint("configuration",
+                                              "config_poll_interval",
+                                              fallback=3600)
+        self._instance_id = Settings.get("configuration", "instance_id")
```
Comment (f1401martin) [inline manager.py]: `self._instance_id` Doesn't seem to be used anywhere...

File: manager.py (hunk 3 above)
Comment (f1401martin) [reply manager.py]: Once this setting is persisted, the only way to override it is through the API, I wish there was a better solution, I wonder if we should have a setting in nio.conf set by default to `False` allowing to override with the nio.conf setting

File: manager.py (hunk 8 above)
Comment (tlugger) [reply manager.py]: I agree

**Pull request merged**